src/librustc/ty/inhabitedness.rs

```diff
@@ -0,0 +1,264 @@
+            where I: IntoIterator<Item=DefIdForrest>
+    {
+        let mut ret = DefIdForrest::empty();
+        let mut next_ret = SmallVec::new();
+        for next_forrest in iter {
+            for id in ret.root_ids.drain(..) {
+                if !next_forrest.contains(tcx, id) {
+                    next_ret.push(id);
+                }
+            }
+
+            for id in next_forrest.root_ids {
+                if !next_ret.contains(&id) {
+                    next_ret.push(id);
+                }
+            }
+
+            mem::swap(&mut next_ret, &mut ret.root_ids);
+            next_ret.drain(..);
+        }
+        ret
+    }
+}
+
+impl<'a, 'gcx, 'tcx> AdtDef {
+    /// Calculate the forrest of DefIds from which this adt is visibly uninhabited.
+    pub fn uninhabited_from(
+                &self,
+                visited: &mut FxHashSet<(DefId, &'tcx Substs<'tcx>)>,
+                tcx: TyCtxt<'a, 'gcx, 'tcx>,
+                substs: &'tcx Substs<'tcx>) -> DefIdForrest
+    {
+        if !visited.insert((self.did, substs)) {
+            return DefIdForrest::empty();
+        }
+
+        let ret = DefIdForrest::intersection(tcx, self.variants.iter().map(|v| {
+            v.uninhabited_from(visited, tcx, substs, self.adt_kind())
+        }));
+        visited.remove(&(self.did, substs));
+        ret
+    }
+}
+
+impl<'a, 'gcx, 'tcx> VariantDef {
+    /// Calculate the forrest of DefIds from which this variant is visibly uninhabited.
+    pub fn uninhabited_from(
+                &self,
+                visited: &mut FxHashSet<(DefId, &'tcx Substs<'tcx>)>,
+                tcx: TyCtxt<'a, 'gcx, 'tcx>,
+                substs: &'tcx Substs<'tcx>,
+                adt_kind: AdtKind) -> DefIdForrest
+    {
+        match adt_kind {
+            AdtKind::Union => {
+                DefIdForrest::intersection(tcx, self.fields.iter().map(|f| {
+                    f.uninhabited_from(visited, tcx, substs, false)
+                }))
+            },
+            AdtKind::Struct => {
+                DefIdForrest::union(tcx, self.fields.iter().map(|f| {
+                    f.uninhabited_from(visited, tcx, substs, false)
+                }))
+            },
+            AdtKind::Enum => {
+                DefIdForrest::union(tcx, self.fields.iter().map(|f| {
+                    f.uninhabited_from(visited, tcx, substs, true)
+                }))
+            },
+        }
+    }
+}
+
+impl<'a, 'gcx, 'tcx> FieldDef {
+    /// Calculate the forrest of DefIds from which this field is visibly uninhabited.
+    pub fn uninhabited_from(
+                &self,
+                visited: &mut FxHashSet<(DefId, &'tcx Substs<'tcx>)>,
+                tcx: TyCtxt<'a, 'gcx, 'tcx>,
+                substs: &'tcx Substs<'tcx>,
+                is_enum: bool) -> DefIdForrest
+    {
+        let mut data_uninhabitedness = move || self.ty(tcx, substs).uninhabited_from(visited, tcx);
+        if is_enum {
```

Comment:
This deserves a comment. Why do we care if this is an enum? Because field privacy is inherited for enums? I think we give a suitable `self.vis` nonetheless (that is, if the enum is public, then `self.vis` will be `Public`) -- I'd prefer to remove this `if`, which seems good esp. if we allow `pub(restricted)` on specific variants or fields of variants someday.